wrappers/vroom.rb

```diff
@@ -29,175 +29,329 @@ def initialize(hash = {})
+        departure_time: begin_time && (begin_time + step['service']),
+        detail: build_detail(service, service.activity, point, nil, vehicle)
+      }.merge(route_data).delete_if{ |_k, v| v.nil? }
+      @previous = point
+      job_data
+    end
+
+    def read_shipment(vrp, vehicle, step)
+      shipment = vrp.shipments[((step['id'] - vrp.services.size) / 2).floor]
+      type = ((step['id'] - vrp.services.size) % 2).zero? ? 'pickup' : 'delivery'
+      activity = type == 'pickup' ? shipment.pickup : shipment.delivery
+      point = activity.point
+      route_data = compute_route_data(vrp, point, step)
+      begin_time = step['arrival'] + step['waiting_time']
+      job_data = {
+        original_shipment_id: shipment.original_id,
+        pickup_shipment_id: type == 'pickup' && shipment.id,
+        delivery_shipment_id: type == 'delivery' && shipment.id,
+        type: type,
+        point_id: point.id,
+        begin_time: begin_time,
+        end_time: begin_time + step['service'],
+        departure_time: begin_time + step['service'],
+        detail: build_detail(shipment, activity, point, nil, vehicle)
+      }.merge(route_data).delete_if{ |_k, v| v.nil? || v == false }
+      @previous = point
+      job_data
+    end
+
+    def compute_route_data(vrp, point, step)
+      return {} if step['type'].nil?
+
+      {
+        travel_time: (@previous && point.matrix_index && vrp.matrices[0][:time] ? vrp.matrices[0][:time][@previous.matrix_index][point.matrix_index] : 0),
+        travel_distance: (@previous && point.matrix_index && vrp.matrices[0][:distance] ? vrp.matrices[0][:distance][@previous.matrix_index][point.matrix_index] : 0),
+        travel_value: (@previous && point.matrix_index && vrp.matrices[0][:value] ? vrp.matrices[0][:value][@previous.matrix_index][point.matrix_index] : 0)
+      }
+    end
+
+    def build_rest(rest)
+      build_detail(nil, rest, nil, nil, nil)
+    end
+
+    def build_detail(_job, activity, point, _day_index, vehicle)
+      {
+        lat: point&.location&.lat,
+        lon: point&.location&.lon,
+        duration: activity&.duration,
+        router_mode: vehicle&.router_mode,
+        speed_multiplier: vehicle&.speed_multiplier
+      }.delete_if{ |_k, v| v.nil? }
+    end
+
+    def run_vroom(vrp, points, dimensions, _job)
       input = Tempfile.new('optimize-vroom-input', @tmp_dir)
       problem = { vehicles: [], jobs: [], matrix: [] }
-      vehicle = vehicles.first
-      problem[:vehicles] << {
-        id: 0,
-        start_index: vehicle.start_point_id ? points[vehicle.start_point_id].matrix_index : nil,
-        end_index: vehicle.end_point_id ? points[vehicle.end_point_id].matrix_index : nil
-      }.delete_if{ |_k, v| v.nil? }
-      problem[:jobs] = services.collect.with_index{ |service, index|
-        [{
+      vrp_skills = vrp.vehicles.flat_map{ |vehicle| vehicle.skills.first }.uniq + vrp.services.flat_map{ |service| service.sticky_vehicles.map(&:id) }.uniq
```

Comment:
If we have `sticky_vehicles`, we should have `sticky_vehicle_ids` as well right?

A note "WARNING: only first alternative set of skills is used" might be useful.

Don't we need uniq one more time in case vehicles have their ID's as skills?

Reply:
`vehicle['sticky_vehicle_ids']` is defined not `vehicle.sticky_vehicle_ids` 

Are we sure that sticky_vehicle_ids doesn't contains duplicated keys ?

Reply:
I know that's what I meant but I have no idea about your question. In that case, we would have the same vehicle in the `sticky_vehicles` list as well, no?

Reply:
It was just an idea by the way, if you think it is risky, no reason to hold the PR, you can mark this resolved.